### ISSUE TITLE
no member named 'ruby_nonempty_memcpy' with Mac 12.1

### ISSUE THREAD

**Issue by ankane** [jasonroelofs/rice#164]
Hi, I'm seeing the following error with Rice and Ruby 3.0.3 on Mac 12.1:

```text
In file included from /Users/andrew/.rbenv/versions/3.0.3/lib/ruby/gems/3.0.0/gems/rice-4.0.2/include/rice/rice.hpp:43:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/ostream:137:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/ios:215:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__locale:15:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/string:511:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/string_view:179:
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__string:388:56: error: no member named 'ruby_nonempty_memcpy' in namespace 'std::__1'; did you mean simply 'ruby_nonempty_memcpy'?
                       : __n == 0 ? __s1 : (char_type*)_VSTD::memcpy(__s1, __s2, __n);
                                                       ^~~~~~~
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__config:856:15: note: expanded from macro '_VSTD'
#define _VSTD std::_LIBCPP_ABI_NAMESPACE
              ^
/Users/andrew/.rbenv/versions/3.0.3/include/ruby-3.0.0/ruby/internal/memory.h:266:1: note: 'ruby_nonempty_memcpy' declared here
ruby_nonempty_memcpy(void *dest, const void *src, size_t n)
^
1 error generated.
make: *** [ext.o] Error 1
```

<strike>I think it may be related to this [bug report](https://bugs.ruby-lang.org/issues/17788) and PR: https://github.com/ruby/ruby/pull/4429</strike>

Edit: Seeing the same issue with Ruby 3.0.2 on Mac 12.1, so think it's related to the new Mac/clang version.

**Comment by ankane** [comment]
It looks like it happens when `<vector>` is included before Rice.

Error

```cpp
#include <vector>
#include <rice/rice.hpp>
```

No error

```cpp
#include <rice/rice.hpp>
#include <vector>
```

**Comment by jasonroelofs** [comment]
Interesting. Does seems like a Ruby bug but will try to pull some time together to take a look.

**Comment by jasonroelofs** [comment]
I'm not able to reproduce this

```
jroelofs@Marvin:~/Development/rice (master)
] ruby -v
ruby 3.0.3p157 (2021-11-24 revision 3fb7d2cadc) [x86_64-darwin20]
jroelofs@Marvin:~/Development/rice (master)
] clang -v
Apple clang version 13.0.0 (clang-1300.0.29.30)
Target: x86_64-apple-darwin21.2.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

**Comment by ankane** [comment]
Hey @jasonroelofs, thanks for looking into it. Here's [a branch](https://github.com/jasonroelofs/rice/compare/master...ankane:debug) that reproduces it on [GitHub Actions](https://github.com/ankane/rice/runs/4831091185?check_suite_focus=true).

**Comment by jasonroelofs** [comment]
Ah, I misunderstood and thought it was a bug manifesting itself in Rice default. I can confirm that if I add `<vector>` like you do I get the same error.

**Comment by jasonroelofs** [comment]
Following the GH discussion (https://github.com/ruby/ruby/pull/4429) I'm not sure there's much I can do to fix this. The fix is applied for Ruby 3.1 but will need an explicit backfill request to get it into Ruby 3.0.

**Comment by ankane** [comment]
Thanks for looking into it. Will close this out since it's not an issue with Rice.
